File: examples/sdk/python/codegen.py
#!/usr/bin/env python3
"""Code generation: decompose, architect, generate, verify, learn.

    decompose ──► architecture ──► gen-<unit> ... (a DAG built from the
        ▲                           │   decomposition; units run in parallel
        │                           ▼   and receive their upstream reports)
        │                         review
        │                           │
        │          verify command ◄─┴─► fix ──► verify ... (bounded)
        └── lessons from earlier runs ◄── retro

The graph is not known until the decomposer replies, so the script builds the
`Workflow` from that reply. Verification is a command *you* pass with
`--verify`; it runs on the host, outside any agent, and its exit status is the
only thing that counts as passing. Generators and the fixer may write and run
commands inside `--workspace`; every other role is denied.

    OPENAI_API_KEY=... python3 codegen.py "A CLI that prints a greeting" \\
        --workspace /tmp/greeter --verify "python3 main.py Ada | grep -q 'Hello, Ada!'"
"""

from __future__ import annotations

import argparse
import asyncio
import json
import re
import subprocess
from pathlib import Path, PurePosixPath
from typing import Any

import tny
from _common import (
    Ledger,
    Lessons,
    Roles,
    add_runtime_arguments,
    ask,
    ask_json,
    log,
    log_event,
    permission_policy,
    require,
    run,
    string_list,
    text_of,
    validate_lessons,
)

_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_-]{0,31}$")


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter
    )
    parser.add_argument("spec", help="the specification, or @FILE to read it from")
    parser.add_argument(
        "--verify", help="shell command that must exit 0, run in --workspace"
    )
    parser.add_argument(
        "--verify-timeout", type=int, default=600, help="seconds (default: 600)"
    )
    parser.add_argument(
        "--max-repairs", type=int, default=2, help="fix attempts (default: 2)"
    )
    parser.add_argument(
        "--max-units", type=int, default=6, help="work units (default: 6)"
    )
    add_runtime_arguments(parser)
    parser.set_defaults(workspace="codegen-out")
    return parser.parse_args()


def validate_units(limit: int) -> Any:
    def validate(value: Any) -> list[dict[str, Any]]:
        require(isinstance(value, dict), "reply must be a JSON object")
        raw = value.get("units")
        require(isinstance(raw, list) and bool(raw), "units must be a non-empty list")
        require(len(raw) <= limit, f"at most {limit} units are allowed")
        units: list[dict[str, Any]] = []
        owners: dict[str, str] = {}
        for item in raw:
            require(isinstance(item, dict), "each unit must be an object")
            identifier, goal = item.get("id"), item.get("goal")
            require(
                isinstance(identifier, str) and bool(_ID.match(identifier)),
                "each id must be a short slug of letters, digits, '-' or '_'",
            )
            require(isinstance(goal, str) and bool(goal.strip()), "goal must be text")
            require(all(identifier != u["id"] for u in units), "ids must be unique")
            files = string_list(item.get("files"), f"{identifier}.files", limit=64)
            require(bool(files), f"unit {identifier} must own at least one file")
            for name in files:
                path = PurePosixPath(name)
                require(
                    not path.is_absolute() and ".." not in path.parts,
                    f"{name} must be a relative path inside the workspace",
                )
                # Parallel generators share one workspace, so file ownership
                # has to be disjoint; the decomposer repairs an overlap.
                require(
                    name not in owners,
                    f"{name} is owned by both {owners.get(name)} and {identifier}",
                )
                owners[name] = identifier
            acceptance = item.get("acceptance")
            units.append(
                {
                    "id": identifier,
                    "goal": goal.strip(),
                    "files": files,
                    "depends_on": string_list(
                        item.get("depends_on", []), f"{identifier}.depends_on", limit=16
                    ),
                    "acceptance": acceptance.strip()
                    if isinstance(acceptance, str)
                    else "",
                }
            )
        known = {unit["id"] for unit in units}
        for unit in units:
            for dependency in unit["depends_on"]:
                require(
                    dependency in known and dependency != unit["id"],
                    f"{unit['id']} depends on unknown unit {dependency}",
                )
        # Report a cycle to the decomposer rather than letting the workflow
        # reject the graph after the model's turn is over.
        done: set[str] = set()
        while len(done) < len(units):
            ready = [
                u["id"]
                for u in units
                if u["id"] not in done and set(u["depends_on"]) <= done
            ]
            require(bool(ready), "unit dependencies contain a cycle")
            done.update(ready)
        return units

    return validate


def build_workflow(
    roles: Roles, spec: str, units: list[dict[str, Any]], advice: str, jobs: int
) -> tny.Workflow:
    workflow = tny.Workflow(
        roles.config("codegen-generator", mode="auto"),
        max_concurrency=jobs,
        on_event=log_event,
        # `auto` already allows edits inside the workspace; this additionally
        # lets generators run their checks. Architect and reviewer override
        # the runtime below and stay read-only.
        on_permission=permission_policy(lambda task: task.startswith("gen-")),
    )
    workflow.task(
        "architecture",
        f"Specification:\n{spec}\n\nWork units:\n{json.dumps(units, indent=2)}{advice}",
        runtime_config=roles.config("codegen-architect"),
    )
    for unit in units:
        acceptance = f"\nAcceptance: {unit['acceptance']}" if unit["acceptance"] else ""
        workflow.task(
            f"gen-{unit['id']}",
            f"Specification:\n{spec}\n\nUnit `{unit['id']}`: {unit['goal']}\n"
            f"Files you own: {', '.join(unit['files'])}{acceptance}",
            depends_on=[
                "architecture",
                *(f"gen-{dependency}" for dependency in unit["depends_on"]),
            ],
        )
    workflow.task(
        "review",
        "Review the generated code in this workspace against the specification "
        "and the architecture contract. The generator reports below are claims "
        f"to check, not facts.\n\nSpecification:\n{spec}",
        depends_on=["architecture", *(f"gen-{unit['id']}" for unit in units)],
        runtime_config=roles.config("review"),  # built-in preset
    )
    return workflow


def run_verify(command: str, workspace: Path, timeout: int) -> tuple[bool, str]:
    try:
        done = subprocess.run(
            command,
            shell=True,
            cwd=workspace,
            capture_output=True,
            text=True,
            errors="replace",
            timeout=timeout,
            check=False,
        )
    except subprocess.TimeoutExpired:
        return False, f"timed out after {timeout}s"
    output = (done.stdout + done.stderr).strip()
    return done.returncode == 0, f"exit {done.returncode}\n{output[-4000:]}"


async def main() -> int:
    args = parse_arguments()
    spec = (
        Path(args.spec[1:]).read_text(encoding="utf-8")
        if args.spec.startswith("@")
        else args.spec
    )
    workspace = Path(args.workspace).resolve()
    workspace.mkdir(parents=True, exist_ok=True)
    roles, ledger = Roles(args), Ledger()
    lessons = Lessons.for_workflow(args, "codegen")
    recalled = lessons.load()
    log(
        f"workspace {workspace}; recalled {len(recalled)} lesson(s) from {lessons.path}"
    )

    units = await ask_json(
        "decompose",
        roles.config("codegen-decomposer"),
        f"Specification:\n{spec}\n\nUse at most {args.max_units} units. JSON shape:\n"
        '{"units": [{"id": "short-slug", "goal": "...", "files": ["relative/path"], '
        '"depends_on": ["other-id"], "acceptance": "..."}]}' + lessons.prompt_block(),
        ledger,
        validate_units(args.max_units),
    )
    log(f"generating {len(units)} unit(s): {', '.join(unit['id'] for unit in units)}")
    for task, role in (
        ("architecture", "codegen-architect"),
        ("gen-*", "codegen-generator"),
        ("review", "review"),
    ):
        log(f"  [{task}] {roles.describe(role)}")

    workflow = build_workflow(roles, spec, units, lessons.prompt_block(), args.jobs)
    result = await workflow.run_async()
    ledger.add_workflow(result)
    statuses = {name: task.status.value for name, task in result.items()}
    for name, task in result.items():
        if not task.ok:
            blocked = f" by {', '.join(task.blocked_by)}" if task.blocked_by else ""
            log(f"  [{name}] {task.status.value}{blocked}: {task.error or ''}")
    review = text_of(result["review"].output) if result["review"].ok else ""

    attempts: list[dict[str, Any]] = []
    passed = result.ok
    while args.verify:
        passed, output = await asyncio.to_thread(
            run_verify, args.verify, workspace, args.verify_timeout
        )
        attempts.append({"passed": passed, "output_tail": output[-1500:]})
        log(f"verify attempt {len(attempts)}: {'passed' if passed else 'FAILED'}")
        if passed or len(attempts) > args.max_repairs:
            break
        # The first repair runs on the fixer's usual tier; a failure that
        # survives it is a critical path and gets the strongest model.
        tier = "critical" if len(attempts) > 1 else None
        await ask(
            f"fix-{len(attempts)}",
            roles.config("codegen-fixer", mode="auto", tier=tier),
            f"Verification command: {args.verify}\n\nIts output:\n{output}\n\n"
            f"Specification:\n{spec}\n\nReviewer findings:\n{review or '(none)'}",
            ledger,
            allow=True,
        )

    learned = await ask_json(
        "retro",
        roles.config("codegen-retro"),
        "Run record:\n"
        + json.dumps(
            {
                "units": units,
                "task_status": statuses,
                "review": review[:4000],
                "verify_command": args.verify,
                "verify_attempts": attempts,
            },
            indent=2,
        )
        + f"\n\nLessons already recorded:\n{json.dumps(recalled[-20:], indent=2)}\n\n"
        'JSON shape: {"lessons": ["..."]}',
        ledger,
        validate_lessons,
    )
    added = lessons.append(learned)

    if review:
        print(review)
    verdict = (
        f"verification {'passed' if passed else 'FAILED'} after {len(attempts)} attempt(s)"
        if args.verify
        else "no --verify command given; nothing was verified"
    )
    log(f"{verdict}; learned {len(added)} new lesson(s); usage: {ledger}")
    return 0 if passed else 1


if __name__ == "__main__":
    run(main)

File: examples/sdk/typescript/codegen.ts
/**
 * Code generation: decompose, architect, generate, verify, learn.
 *
 *     decompose ──► architecture ──► gen-<unit> ... (a DAG built from the
 *         ▲                           │   decomposition; units run in parallel
 *         │                           ▼   and receive their upstream reports)
 *         │                         review
 *         │                           │
 *         │          verify command ◄─┴─► fix ──► verify ... (bounded)
 *         └── lessons from earlier runs ◄── retro
 *
 * The graph is not known until the decomposer replies, so the script builds
 * the `Workflow` from that reply. Verification is a command *you* pass with
 * `--verify`; it runs on the host, outside any agent, and its exit status is
 * the only thing that counts as passing. Generators and the fixer may write
 * and run commands inside `--workspace`; every other role is denied.
 *
 *     OPENAI_API_KEY=... node codegen.ts "A CLI that prints a greeting" \
 *         --workspace /tmp/greeter --verify "python3 main.py Ada | grep -q 'Hello, Ada!'"
 */

import { exec } from "node:child_process";
import { mkdirSync, readFileSync } from "node:fs";
import { isAbsolute, resolve } from "node:path";
import { parseArgs, promisify } from "node:util";

import { Workflow } from "@thehumanworks/tny";

import {
  Ledger,
  Lessons,
  RUNTIME_HELP,
  Roles,
  UsageError,
  ask,
  askJson,
  integer,
  isRecord,
  log,
  logEvent,
  permissionPolicy,
  requireReply,
  run,
  runtimeArguments,
  runtimeArgumentsFrom,
  stringList,
  validateLessons,
} from "./common.ts";

const HELP = `usage: node codegen.ts [options] SPEC      (SPEC may be @FILE)

  --verify CMD          shell command that must exit 0, run in --workspace
  --verify-timeout SEC  (default: 600)
  --max-repairs N       fix attempts (default: 2)
  --max-units N         work units (default: 6)
${RUNTIME_HELP}
  (--workspace defaults to ./codegen-out)`;

const ID = /^[A-Za-z0-9][A-Za-z0-9_-]{0,31}$/;

interface Unit {
  id: string;
  goal: string;
  files: string[];
  depends_on: string[];
  acceptance: string;
}

function validateUnits(limit: number): (value: unknown) => Unit[] {
  return (value) => {
    requireReply(isRecord(value), "reply must be a JSON object");
    const raw = value.units;
    requireReply(Array.isArray(raw) && raw.length > 0, "units must be a non-empty list");
    requireReply(raw.length <= limit, `at most ${limit} units are allowed`);
    const units: Unit[] = [];
    const owners = new Map<string, string>();
    for (const item of raw) {
      requireReply(isRecord(item), "each unit must be an object");
      const { id, goal, acceptance } = item;
      requireReply(typeof id === "string" && ID.test(id), "each id must be a short slug of letters, digits, '-' or '_'");
      requireReply(typeof goal === "string" && goal.trim() !== "", "goal must be text");
      requireReply(units.every((unit) => unit.id !== id), "ids must be unique");
      const files = stringList(item.files, `${id}.files`, 64);
      requireReply(files.length > 0, `unit ${id} must own at least one file`);
      for (const name of files) {
        requireReply(
          !isAbsolute(name) && !name.split("/").includes(".."),
          `${name} must be a relative path inside the workspace`,
        );
        // Parallel generators share one workspace, so file ownership has to
        // be disjoint; the decomposer repairs an overlap.
        requireReply(!owners.has(name), `${name} is owned by both ${owners.get(name)} and ${id}`);
        owners.set(name, id);
      }
      units.push({
        id,
        goal: goal.trim(),
        files,
        depends_on: stringList(item.depends_on ?? [], `${id}.depends_on`, 16),
        acceptance: typeof acceptance === "string" ? acceptance.trim() : "",
      });
    }
    const known = new Set(units.map((unit) => unit.id));
    for (const unit of units) {
      for (const dependency of unit.depends_on) {
        requireReply(
          known.has(dependency) && dependency !== unit.id,
          `${unit.id} depends on unknown unit ${dependency}`,
        );
      }
    }
    // Report a cycle to the decomposer rather than letting the workflow
    // reject the graph after the model's turn is over.
    const done = new Set<string>();
    while (done.size < units.length) {
      const ready = units.filter((unit) => !done.has(unit.id) && unit.depends_on.every((id) => done.has(id)));
      requireReply(ready.length > 0, "unit dependencies contain a cycle");
      for (const unit of ready) done.add(unit.id);
    }
    return units;
  };
}

function buildWorkflow(roles: Roles, spec: string, units: readonly Unit[], advice: string, jobs: number): Workflow {
  const workflow = new Workflow({
    runtime: roles.options("codegen-generator", "auto"),
    maxConcurrency: jobs,
    onEvent: logEvent,
    // `auto` already allows edits inside the workspace; this additionally
    // lets generators run their checks. Architect and reviewer override the
    // runtime below and stay read-only.
    onPermission: permissionPolicy((task) => task.startsWith("gen-")),
  });
  workflow.task(
    "architecture",
    `Specification:\n${spec}\n\nWork units:\n${JSON.stringify(units, null, 2)}${advice}`,
    { runtime: roles.options("codegen-architect") },
  );
  for (const unit of units) {
    const acceptance = unit.acceptance ? `\nAcceptance: ${unit.acceptance}` : "";
    workflow.task(
      `gen-${unit.id}`,
      `Specification:\n${spec}\n\nUnit \`${unit.id}\`: ${unit.goal}\n` +
        `Files you own: ${unit.files.join(", ")}${acceptance}`,
      { dependsOn: ["architecture", ...unit.depends_on.map((id) => `gen-${id}`)] },
    );
  }
  workflow.task(
    "review",
    "Review the generated code in this workspace against the specification " +
      "and the architecture contract. The generator reports below are claims " +
      `to check, not facts.\n\nSpecification:\n${spec}`,
    {
      dependsOn: ["architecture", ...units.map((unit) => `gen-${unit.id}`)],
      runtime: roles.options("review"), // built-in preset
    },
  );
  return workflow;
}

async function runVerify(command: string, workspace: string, timeoutSeconds: number): Promise<[boolean, string]> {
  try {
    const { stdout, stderr } = await promisify(exec)(command, {
      cwd: workspace,
      timeout: timeoutSeconds * 1000,
      maxBuffer: 16 * 1024 * 1024,
    });
    return [true, `exit 0\n${(stdout + stderr).trim().slice(-4000)}`];
  } catch (error) {
    const failed = error as { code?: number | string; killed?: boolean; stdout?: string; stderr?: string };
    if (failed.killed) return [false, `timed out after ${timeoutSeconds}s`];
    const output = `${failed.stdout ?? ""}${failed.stderr ?? ""}`.trim();
    return [false, `exit ${failed.code ?? "unknown"}\n${output.slice(-4000)}`];
  }
}

async function main(): Promise<number> {
  const { values, positionals } = parseArgs({
    allowPositionals: true,
    options: {
      verify: { type: "string" },
      "verify-timeout": { type: "string", default: "600" },
      "max-repairs": { type: "string", default: "2" },
      "max-units": { type: "string", default: "6" },
      ...runtimeArguments,
    },
  });
  if (values.help) {
    console.log(HELP);
    return 0;
  }
  const specArgument = positionals[0];
  if (specArgument === undefined || positionals.length !== 1) throw new UsageError(`expected exactly one SPEC\n${HELP}`);
  const spec = specArgument.startsWith("@") ? readFileSync(specArgument.slice(1), "utf8") : specArgument;
  const verifyTimeout = integer(values["verify-timeout"], "--verify-timeout", 1);
  const maxRepairs = integer(values["max-repairs"], "--max-repairs", 0);
  const maxUnits = integer(values["max-units"], "--max-units", 1);

  const args = runtimeArgumentsFrom(values, "codegen-out");
  const workspace = resolve(args.workspace);
  mkdirSync(workspace, { recursive: true });
  const roles = new Roles(args);
  const ledger = new Ledger();
  const lessons = Lessons.forWorkflow(args, "codegen");
  const recalled = lessons.load();
  log(`workspace ${workspace}; recalled ${recalled.length} lesson(s) from ${lessons.path}`);

  const units = await askJson(
    "decompose",
    roles.options("codegen-decomposer"),
    `Specification:\n${spec}\n\nUse at most ${maxUnits} units. JSON shape:\n` +
      '{"units": [{"id": "short-slug", "goal": "...", "files": ["relative/path"], ' +
      '"depends_on": ["other-id"], "acceptance": "..."}]}' +
      lessons.promptBlock(),
    ledger,
    validateUnits(maxUnits),
  );
  log(`generating ${units.length} unit(s): ${units.map((unit) => unit.id).join(", ")}`);
  for (const [task, role] of [
    ["architecture", "codegen-architect"],
    ["gen-*", "codegen-generator"],
    ["review", "review"],
  ] as const) {
    log(`  [${task}] ${roles.describe(role)}`);
  }

  const result = await buildWorkflow(roles, spec, units, lessons.promptBlock(), args.jobs).run();
  ledger.addWorkflow(result);
  for (const task of result.results) {
    if (task.ok) continue;
    const blocked = task.blockedBy.length > 0 ? ` by ${task.blockedBy.join(", ")}` : "";
    log(`  [${task.name}] ${task.status}${blocked}: ${task.error?.message ?? ""}`);
  }
  const reviewed = result.require("review");
  const review = reviewed.ok ? reviewed.output : "";

  const attempts: { passed: boolean; output_tail: string }[] = [];
  let passed = result.ok;
  while (values.verify) {
    let output: string;
    [passed, output] = await runVerify(values.verify, workspace, verifyTimeout);
    attempts.push({ passed, output_tail: output.slice(-1500) });
    log(`verify attempt ${attempts.length}: ${passed ? "passed" : "FAILED"}`);
    if (passed || attempts.length > maxRepairs) break;
    // The first repair runs on the fixer's usual tier; a failure that
    // survives it is a critical path and gets the strongest model.
    const tier = attempts.length > 1 ? "critical" : undefined;
    await ask(
      `fix-${attempts.length}`,
      roles.options("codegen-fixer", "auto", tier),
      `Verification command: ${values.verify}\n\nIts output:\n${output}\n\n` +
        `Specification:\n${spec}\n\nReviewer findings:\n${review || "(none)"}`,
      ledger,
      true,
    );
  }

  const record = {
    units,
    task_status: Object.fromEntries(result.results.map((task) => [task.name, task.status])),
    review: review.slice(0, 4000),
    verify_command: values.verify ?? null,
    verify_attempts: attempts,
  };
  const learned = await askJson(
    "retro",
    roles.options("codegen-retro"),
    `Run record:\n${JSON.stringify(record, null, 2)}` +
      `\n\nLessons already recorded:\n${JSON.stringify(recalled.slice(-20), null, 2)}\n\n` +
      'JSON shape: {"lessons": ["..."]}',
    ledger,
    validateLessons,
  );
  const added = lessons.append(learned);

  if (review) console.log(review);
  const verdict = values.verify
    ? `verification ${passed ? "passed" : "FAILED"} after ${attempts.length} attempt(s)`
    : "no --verify command given; nothing was verified";
  log(`${verdict}; learned ${added.length} new lesson(s); usage: ${ledger}`);
  return passed ? 0 : 1;
}

await run(main);
